### DISCUSSION TITLE
Allow Scarpet apps to specify a minimum mod version

### DISCUSSION THREAD

**Discussion by altrisi** [gnembon/fabric-carpet#644] [Ideas]
_Moved from #642, mostly to test this discussions thing_

Currently, there is no way for an app to ask for a minimum Carpet/extension version, and if it's using newer features, it will just fail with something similar to `Function not defined`, or just silently failing to apply some new feature or event, which could derive to app malfunction.

Therefore I suggest adding a new map to the `__config()` to prevent an app from loading in case the current Carpet version doesn't apply to the requirements set by the app.

It could be something like the following:

```py

__config() -> 
{
  'requires' -> 
  {
    'carpet' -> '>=1.14.20',
    'some-scarpet-extension' -> '1.2.3',
    'yet-another-extension' -> '*'
  }
}
```

It could be based on the modid, and then use FabricLoader's (not Fabric API) version "APIs" (it is located in the `api` package) to check for compatibility directly using mod's metadata parser, so it would allow all operators supported by Fabric in Scarpet. That could be useful, for example, in case it just needs the extension to be present, or if it's using old, removed functions and no one is willing to update it, it could also specify, for instance `'carpet' -> '<=1.14.19'`.

_This could make users not be able to load an app even if it's compatible_: Well, actually not, since if they know it's compatible, they also probably know how to delete that from the config. And also, this would tell people to update, which I don't think is bad.

Thoughts?

**Comment by Firigion** [comment]
Sounds interesting. Could also add a game verion check to that. I know you can do it manualy, but getting a message that says "Failed to load" seems better than just having to test by hand.

**Reply by altrisi** [reply]
Actually, if this was implemented using Loader APIs, Minecraft is considered a "mod", so it would work out of the box.

**Comment by replaceitem** [comment]
Thats a good idea, why not

**Comment by gnembon** [comment]
technically you could just type in at the top of the script in the static section
if (system_info('scarpet_version') < .....,
    exit('Requires 1.7.8')
);
for example - that will even prevent the config from loading

**Comment by Ghoulboy78** [comment]
I actually went a ways into implementing this

**Reply by Ghoulboy78** [reply]
https://github.com/Ghoulboy78/fabric-carpet/tree/shebang-lines(sorta)

**Reply by Ghoulboy78** [reply]
I also added a way for carpet extensions to use the version integer (if they want to) to help with this, though of course this would require a pr to gnembon/fabric-carpet-example-extesion-mod or whatevs. But that is unlikely to happen, and it's more likely that htye just update the string. Therefore I use regex to check if they have `MAJOR_VERSION.MINOR_VERSION.PATCH_VERSION` format, and use that to check. But its stilll not finished or tested, so I havent opened a pr.

**Reply by altrisi** [reply]
U should have used FLoader's version checking, so you could check any mod (not only Carpet and extensions), wouldn't need extensions to declare themselves (most don't even know those methods exist), and it has semver parsing and comparison already integrated.
"Mod" even includes MC version in FLoader's terms.

**Reply by altrisi** [reply]
BTW isn't that branch a bit old?

**Reply by Ghoulboy78** [reply]
It is kinda, and I didn't know about any FLoader, so idk.

**Reply by altrisi** [reply]
I mentioned it on the top in case someone wanted to try though, but idk, I guess this thread got a bit old.

**Reply by Ghoulboy78** [reply]
I hadnt read this tread. I only found out that this repo had GitHyb discussions like today

**Reply by altrisi** [reply]
Oh then nvm.

**Comment by altrisi** [comment]
Just sent a PR with the implementation I started back in the days of this discussion as #798.

Would appreciate feedback about how it works here, and about the actual implementation there (or wherever you like really).

**Reply by altrisi** [reply]
Implemented in latest Carpet release! :tada:
